### PR TITLE
🏗  ♻️  Fix ALMOST all type issues in /build-system

### PR DESCRIPTION
**Pull request by rileyajones**
Prior to this change there were 99 type errors, assuming no more are introduce prior to this being merged there will only be one after. The remaining issue is quite complicated and I plan to fix it as a separate PR.

#28387

### CODE REVIEW

File: build-system/compile/post-closure-babel.js [hunk 1]
```diff
@@ -84,9 +84,10 @@ async function postClosureBabel(file) {
   }
 
   debug(CompilationLifecycles['closured-pre-babel'], file);
+  /** @type {?babel.TransformOptions} */
   const babelOptions = babel.loadOptions({caller: {name: 'post-closure'}});
   const {code, map: babelMap} =
-    (await babel.transformFileAsync(file, babelOptions)) || {};
+    (await babel.transformFileAsync(file, babelOptions ?? undefined)) || {};
```
Comment (rileyajones) [inline build-system/compile/post-closure-babel.js:90]: `babel.loadOptions` returns `babel.TransformOptions | null` while `babel.transformFileAsync` only accepts `babel.TransformOptions | undefined`

File: build-system/tasks/release/index.js [hunk 2]
```diff
@@ -125,9 +136,10 @@ async function prepareEnvironment_(outputDir, tempDir) {
  * @return {!Array<!Object>} list of AMP flavors to build.
  */
 function discoverDistFlavors_() {
+  const experimentConfigDefs = /** @type {[string, ExperimentConfigDef][]} */ (Object.entries(experimentsConfig));
   const distFlavors = [
     BASE_FLAVOR_CONFIG,
-    ...Object.entries(experimentsConfig)
+    ...experimentConfigDefs
```
Comment (rileyajones) [inline build-system/tasks/release/index.js:142]: Breaking out `experimentConfigDefs` into a variable allows a cleaner type assertion.

File: build-system/tasks/e2e/describes-e2e.js [hunk 3]
```diff
@@ -490,7 +485,6 @@ function describeEnv(factory) {
         // If there is an async expect error, throw it in the final state.
         const lastExpectError = getLastExpectError();
         if (lastExpectError) {
-          this.test.error(lastExpectError);
```
Comment (rileyajones) [inline build-system/tasks/e2e/describes-e2e.js:493]: `Mocha.Runnable.error` is not a function and unless I am missing something we never extend it so I do not believe this function actually exists. Please lmk if I am missing some thing here.

File: build-system/compile/compile.js [hunk 4]
```diff
@@ -50,6 +50,9 @@ const MAX_PARALLEL_CLOSURE_INVOCATIONS =
  *  typeCheckOnly?: boolean,
  *  skipUnknownDepsCheck?: boolean,
  *  warningLevel?: boolean,
+ *  noAddDeps?: boolean,
+ *  continueOnError?: boolean,
+ *   errored?: boolean,
```
Comment (rsimha) [inline build-system/compile/compile.js:55]: nit: alignment
```suggestion
 *  errored?: boolean,
```

File: build-system/compile/compile.js [hunk 5]
```diff
@@ -120,13 +123,11 @@ function cleanupBuildDir() {
...
  *
- * @param {string[]|string} entryModuleFilenames
- * @param {string} outputDir
- * @param {string} outputFilename
+ * @param {string[]} entryModuleFilenames
  * @param {!OptionsDef} options
  * @return {!Array<string>}
  */
-function getSrcs(entryModuleFilenames, outputDir, outputFilename, options) {
+function getSrcs(entryModuleFilenames, options) {
```
Comment (rsimha) [inline build-system/compile/compile.js:130]: More proof that type-checking does help.

File: build-system/compile/debug-compilation-lifecycle.js [hunk 6]
```diff
@@ -42,7 +42,7 @@ const LIFECYCLES = {
  *
  * @param {string} lifecycle
  * @param {string} fullpath
- * @param {string=} content
+ * @param {?string=} content
```
Comment (rsimha) [inline build-system/compile/debug-compilation-lifecycle.js:45]: This seems wrong because the JSDoc says the last two params have default values, but the function signature doesn't specify any.

I think it's best to make the last two params purely optional and drop the default specifier.

File: build-system/compile/post-closure-babel.js (hunk 1 above)
Comment (rsimha) [reply build-system/compile/post-closure-babel.js:90]: Weird mismatch by babel, but this fix works.

File: build-system/compile/post-closure-babel.js [hunk 7]
```diff
@@ -101,8 +102,12 @@ async function postClosureBabel(file) {
     () => null,
     !argv.full_sourcemaps
   );
-
-  debug(CompilationLifecycles['complete'], file, compressed, sourceMap);
+  debug(
+    CompilationLifecycles['complete'],
+    file,
+    /** @type {string|undefined} */ (compressed),
```
Comment (rsimha) [inline build-system/compile/post-closure-babel.js:108]: Is there a better way to do this? What is the original type of `compressed`? Can we just pass in `compressed?.toString()` here and ditch the cast?

File: build-system/server/new-server/router.js [hunk 8]
```diff
@@ -15,6 +15,7 @@
  */
 
 const router = require('express').Router();
+// @ts-ignore
```
Comment (rsimha) [inline build-system/server/new-server/router.js:18]: Totally good with this. The stuff in `dist/transform` is output that we shouldn't be type-checking. (I assume adding this directory to the ignore list isn't sufficient?)

File: build-system/tasks/coverage-map/index.js [hunk 9]
```diff
@@ -66,8 +66,9 @@ async function collectCoverage() {
 
 /**
  * Source: https://github.com/chenxiaochun/blog/issues/38s
- *
- * @param {puppeteer.Page} page
+ * TODO(28387) change the first parameter to puppeteer.Page once puppeteer is consistently
```
Comment (rsimha) [inline build-system/tasks/coverage-map/index.js:69]: nit: linkify
```suggestion
 * TODO(#28387) change the first parameter to puppeteer.Page once puppeteer is consistently
```

Same comment applies to the rest of the TODOs in this PR.

File: build-system/tasks/e2e/describes-e2e.js (hunk 3 above)
Comment (rsimha) [reply build-system/tasks/e2e/describes-e2e.js:493]: This line was deliberately added in #22366. I don't think it should be removed until we're certain there will be no repercussions. Maybe consult with @estherkim who wrote this code, and until then, merely silence the error?

File: build-system/tasks/extension-helpers.js [hunk 10]
```diff
@@ -91,25 +91,17 @@ const DEFAULT_EXTENSION_SET = ['amp-loader', 'amp-auto-lightbox'];
  *   cssBinaries?: Array<string>,
  *   extraGlobs?: Array<string>,
  *   binaries?: Array<ExtensionBinary>,
- *   npm?: Map<string, NpmBinaries>,
+ *   npm?: boolean,
  * }}
  */
 const ExtensionOption = {}; // eslint-disable-line no-unused-vars
```
Comment (rsimha) [inline build-system/tasks/extension-helpers.js:97]: Pretty sure you can make this `ExtensionOptionDef` and remove the lint override.

File: build-system/tasks/extension-helpers.js [hunk 11]
```diff
@@ -91,25 +91,17 @@ const DEFAULT_EXTENSION_SET = ['amp-loader', 'amp-auto-lightbox'];
...
 /**
  * @typedef {{
  *   entryPoint: string,
  *   outfile: string,
  *   external?: Array<string>
- *   remap?: Map<string, string>
+ *   remap?: Record<string, string>
  * }}
  */
 const ExtensionBinary = {}; // eslint-disable-line no-unused-vars
```
Comment (rsimha) [inline build-system/tasks/extension-helpers.js:107]: Same here, `ExtensionBinaryDef`.

**Review by rsimha — APPROVED** [review]
Bravo! LGTM with a few comments below.

File: build-system/compile/post-closure-babel.js (hunk 7 above)
Comment (rileyajones) [reply build-system/compile/post-closure-babel.js:108]: Yeah, that works

File: build-system/server/new-server/router.js (hunk 8 above)
Comment (rileyajones) [reply build-system/server/new-server/router.js:18]: Adding it to the `tsconfig.excludes` does not stop it from being checked as a dependency. This is here to handle the situation where `new-server` hasn't been built yet and thus `dist` does not exist

File: build-system/tasks/extension-helpers.js (hunk 10 above)
Comment (rileyajones) [reply build-system/tasks/extension-helpers.js:97]: Great point!

File: build-system/compile/compile.js (hunk 5 above)
Comment (rileyajones) [reply build-system/compile/compile.js:130]: Yeah, I think it's really highlighted by the `TODO` on line 190 which shows you thought there needed to be clean up here but it wasn't clear that two of the parameters were not being used.

File: build-system/tasks/e2e/describes-e2e.js (hunk 3 above)
Comment (rileyajones) [reply build-system/tasks/e2e/describes-e2e.js:493]: Alright, I was a little nervous about this one (thus the comment).

Comment (estherkim) [reply build-system/tasks/e2e/describes-e2e.js:493]: Relevant https://github.com/mochajs/mocha/issues/4515

tl;dr: it's not a documented API but works

File: build-system/compile/debug-compilation-lifecycle.js (hunk 6 above)
Comment (rileyajones) [reply build-system/compile/debug-compilation-lifecycle.js:45]: The function signature does not specify a default, but it does assign a default value in the function body. Would this make more sense to you if I moved the default assignment to the signature?

Comment (rsimha) [reply build-system/compile/debug-compilation-lifecycle.js:45]: That wouldn't work because in the event `content` and `sourcemap` are not passed in, this function reads them from a file, and we don't want to have a file read operation in the function signature.

I think the suggestion I posted above is the cleanest way to do it (make the params strictly optional and drop the default). So, `@param {?string}` for both `content` and `sourcemap`. Would that take care of the type errors?

File: build-system/tasks/coverage-map/index.js (hunk 9 above)
Comment (rsimha) [reply build-system/tasks/coverage-map/index.js]: nit: Unresolving because there are a few more instances of `TODO(XXXXX)` that can be linkified with a `#`.

File: build-system/compile/debug-compilation-lifecycle.js (hunk 6 above)
Comment (rileyajones) [reply build-system/compile/debug-compilation-lifecycle.js:45]: I'm a little confused by this suggestion. `@param {?string}` is equivalent to `@param {string|null}` `@param {string=}` is intended to indicate the parameter is optional using closure syntax (https://jsdoc.app/tags-param.html#optional-parameters-and-default-values).

`@param {string=}` will result in a single type error in [esbuild-bable](https://github.com/ampproject/amphtml/blob/main/build-system/common/esbuild-babel.js#L60) where the variable `code` is of type `string | null | undefined`. While I could add an assertion there I think the usage is valid.

`@param {?string}` results in 6 type errors from the places where the function is called without 4 parameters.

File: build-system/tasks/coverage-map/index.js (hunk 9 above)
Comment (rileyajones) [reply build-system/tasks/coverage-map/index.js]: Resolved for real this time

File: build-system/compile/debug-compilation-lifecycle.js (hunk 6 above)
Comment (rsimha) [reply build-system/compile/debug-compilation-lifecycle.js:45]: Aha, I see. Thanks for explaining this to me. I now stand corrected, carry on!

**Review by rsimha — APPROVED** [review]
Final version LGTM. Thanks for ushering this through all the reviews!
